accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+
+Although the writable DOM and `dynamic` features are somewhat separate, considering both during design should result in a single API for both scenarios that is optimized to have minimal state and has intuitive interop between the DOM classes and those returned when using `dynamic`.
+
+The existing `JsonDocument` class represents the DOM support today, however it is read-only. It maintains a single immutable UTF-8 buffer and returns `JsonElement` value types from that buffer on demand, which minimizes the initial `JsonDocument.Parse()` time and heap allocs but also makes it slow to re-obtain the same value (which is common with LINQ) and does not lend itself to being extended to support writability.
+
+This document proposes a new API based on learning and scenarios from:
+- The writable DOM prototype.
+  - During 5.0, there was a [writable DOM effort](https://github.com/dotnet/runtime/issues/30436). Unlike `JsonDocument`, the design centered around writable aspects, usability and LINQ support, and not performance and allocations. It was shelved due to time constraints and outstanding work.
+- The dynamic support code example.
+  - During 5.0, [dynamic support](https://github.com/dotnet/runtime/issues/29690) was not able to be implemented due to schedule constraints. Instead, a [code example](https://github.com/dotnet/runtime/pull/42097) was provided that enables `dynamic` and a writeable DOM that can be used by any 3.0+ runtime in order to unblock the community and gather feedback for a 6.0 feature.
+- Azure prototyping.
+  - The Azure team needs a writable DOM, and supporting `dynamic` is important for some scenarios but not all. A [prototype](https://docs.microsoft.com/en-us/dotnet/api/azure.core.dynamicjson?view=azure-dotnet-preview) has been created. Work is also being coordinated with the C# team to support Intellisense with `dynamic`, although that is not expected to be implemented in time to be used in 6.0.
+
+# Dynamic
+The dynamic feature in C# is enabled by both language features and implementation in the `System.Linq.Expressions` assembly. The core interface detected by the language to enable this support is [IDynamicMetaObjectProvider](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.idynamicmetaobjectprovider?view=net-5.0).
+
+There are implementations of `IDynamicMetaObjectProvider` including [DynamicObject](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.dynamicobject?view=net-5.0) and [ExpandObject](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.expandoobject?view=netcore-3.1). However, both implementations are not ideal:
+- There are many public members used for wiring up dynamic support, but are of little to no value for consumers so they would be confusing when mixed in with other methods intended to be used for writable DOM support.
+- It is not possible to efficiently combine writable DOM classes that are not tied to `dynamic` and `System.Linq.Expressions`:
```

Comment:
The assumes the current implementation of dynamic. What would it take to change the implementation of dynamic to allow this combining to be done efficiently?

Reply:
The goal is to avoid the dependency to SLE.dll for cases that don't actually use `dynamic` at run-time.

Today, even if `IDynamicMetaObjectProvider` is placed into a new small assembly, the one member the interface has expands into many more types in SLE.dll:
```cs
System.Dynamic.DynamicMetaObject GetMetaObject (System.Linq.Expressions.Expression parameter)
```

One approach is to add a new `IDynamicMetaObjectProviderEx` interface in a small assembly (probably with little to no code) and where the equivalent of `GetMetaObject` only returns types from that same small assembly.

Another approach, which may not be feasible, is to expand the linker in some way to handle this.

cc @MadsTorgersen

Reply:
> The goal is to avoid the dependency to SLE.dll for cases that don't actually use dynamic at run-time.

I agree with this.

> Today, even if IDynamicMetaObjectProvider is placed into a new small assembly,

This is just one of the design options. For example, an interesting option may be to introduce a new light-weight interface in System.Runtime that supports dynamic scenarios like this one.